### ISSUE TITLE
create-react-app: TypeError: Class constructor [...] cannot be invoked without 'new'

### ISSUE THREAD

**Issue by sisp** [xaviergonz/mobx-keystone#15]
When using `ExtendedModel` in a `create-react-app`-based app, I'm getting the following error:

> TypeError: Class constructor [...] cannot be invoked without 'new'

This is a minimal reproduction: https://github.com/sisp/mobx-keystone-cra-issue

The error does not occur when I run the example with `ts-node`, so I think the problem originates from CRA's usage of Babel + Typescript build chain.

**Comment by sisp** [comment]
I think there is a relationship with [this TSDX PR](https://github.com/palmerhq/tsdx/pull/111). But two commits later, the [Typescript config override was set back to before the PR](https://github.com/palmerhq/tsdx/commit/b6492a193d8b6d2ae536492aac41690e3f80d0b3#diff-85d99c527c01407b9055b07762a4ad45R153). I think `target: es5` wouldn't work for mobx-keystone anyway because ES6 sets are used, so it couldn't be transpiled to ES5 anyway.

**Comment by sisp** [comment]
I'm getting closer to understanding the problem. Here is a [quote from @gaeron about this problem](https://github.com/facebook/create-react-app/issues/1751#issuecomment-284796527):

> I’m not sure if you read the linked issue, but it’s not extending native classes that’s broken per se, but extending them using Babel transform. If A is native, but B is written with Babel class transform, and B extends A, it will break. This is a known Babel issue (I linked to it), and Babel likely won’t fix it because of performance concerns with the fully compatible solution. This doesn’t come up very often because it works the other way around. So it’s only a problem if a library uses Babel and that library happens to extend your class which might be native.

I think this is relevant here.

**Comment by xaviergonz** [comment]
It depends actually on the target of the user side transpiled code and the library transpiled code
Basically, typescript, when transpiling to ES5 will create classes as functions, with a call to BASE.call(this, ...) to invoke the constructor, but with ES6 class you are forced to use new

There are two ways to fix it without touching the code
1) make the user side code transpile to ES5, so the ES5 library can call apply on the base class
2) create a es6 version of the library that keeps the non transpiled class

Basically, the user side and library must match (either es5 or es6)

Does CRA use the ESM version or the common js one? if so maybe the es version could be defaulted to compile to es6

**Comment by xaviergonz** [comment]
Or maybe it is just the time to ship ES6 by default, but that might keep people with old browsers out :(

**Comment by xaviergonz** [comment]
Basically
- FC (function class) extends FC is ok
- class extends class is ok
- class extends FC is ok
- FC extends class is broken

If the user code is ES6 and the library is ES5 then for a non extended model it will be

class -> FC (Model) -> FC (BaseModel)

but for extended model it is

class -> FC (ExtendedModel) -> class -> FC (Model) -> FC (BaseModel)

A solution keeping es5 is possible btw, but it requires proxies :-/

**Comment by sisp** [comment]
> Does CRA use the ESM version or the common js one?

With #14 merged, CRA uses the ESM version. Before #14, it was using CommonJS.

> if so maybe the es version could be defaulted to compile to es6

Does [the following line from TSDX's Rollup config](https://github.com/palmerhq/tsdx/blob/73b152eca021ab4897efa92d8ec3abb893fc8b96/src/createRollupConfig.ts#L40) mean that TSDX compiles the library to ES5 because `target: undefined` when `opts.target === 'browser'`? 

```ts
targets: opts.target === 'node' ? { node: '8' } : undefined,
```

> There are two ways to fix it without touching the code
>
> 1. make the user side code transpile to ES5, so the ES5 library can call apply on the base class

In the [example I provided to reproduce this error](https://github.com/sisp/mobx-keystone-cra-issue), I believe [Typescript is configured to transpile to ES5](https://github.com/sisp/mobx-keystone-cra-issue/blob/42613d0c8a65dd177ad745d153afac0888d31376/tsconfig.json#L3), isn't it?

```json
"target": "es5",
```

Yet I'm getting the error.

> 2. create a es6 version of the library that keeps the non transpiled class

What would need to be changed for this? By the way, since [TSDX v0.9.0](https://github.com/palmerhq/tsdx/releases/tag/v0.9.0), TSDX's Rollup config can be customized.

> Basically, the user side and library must match (either es5 or es6)
> ...
> A solution keeping es5 is possible btw, but it requires proxies :-/

Is my understanding correct that there is no way to support client code for both ES5 and ES6 simultaneously without ES6 Proxy? You're using ES6 Set (e.g. [here](https://github.com/xaviergonz/mobx-keystone/blob/b047eb39f270afb3b2e3ea04d0ec82a6b27e5d48/packages/lib/src/action/builtInActions.ts#L19)), so doesn't this require the client target to be ES6 anyway? Or is ES6 Set polyfilled?

If there is no way to support ES5 and ES6 client code with the same library, would it make sense to build 2 versions, one for ES5 and one for ES6, and publish them separately? This would be somewhat similar to `lodash` and `lodash-es` I think.

**Comment by xaviergonz** [comment]
> In the example I provided to reproduce this error, I believe Typescript is configured to transpile to ES5, isn't it? "target": "es5", Yet I'm getting the error.

Because CRA ignores that for development mode, if you look at the main.js file generated in dev mode in the chrome inspector you will see this
```
class A extends Object(mobx_keystone__WEBPACK_IMPORTED_MODULE_0__["Model"])({}) {} // Uncomment the below block and it works fine.
//
// @model("B")
// class B extends A {}


let B = (_dec = Object(mobx_keystone__WEBPACK_IMPORTED_MODULE_0__["model"])("B"), _dec(_class = class B extends Object(mobx_keystone__WEBPACK_IMPORTED_MODULE_0__["ExtendedModel"])(A, {}) {}) || _class); // TypeError: Class constructor A cannot be invoked without 'new'
```

Either way I think I found a solution with the PR that works for everything :)

**Comment by sisp** [comment]
> Either way I think I found a solution with the PR that works for everything :)

Indeed, thanks a lot, again! :-)
